R/as.data.table.R

```diff
@@ -64,20 +63,21 @@ as.data.table.matrix <- function(x, keep.rownames=FALSE, ...) {
     for (i in ic) value[[i]] <- as.vector(x[, i])       # to drop any row.names that would otherwise be retained inside every column of the data.table
   }
   col_labels <- dimnames(x)[[2L]]
+  setDT(value)
   if (length(col_labels) == ncols) {
     if (any(empty <- !nzchar(col_labels)))
       col_labels[empty] <- paste0("V", ic[empty])
-    setattr(value, "names", col_labels)
+    setnames(value, col_labels)
   } else {
-    setattr(value, "names", paste0("V", ic))
+    setnames(value, paste0("V", ic))
   }
-  setattr(value,"row.names",.set_row_names(nrows))
-  setattr(value,"class",c("data.table","data.frame"))
-  alloc.col(value)
+  # setkey now to allow matrix column names as key
+  setkeyv(value, key)
+  value
 }
 
 # as.data.table.array - #1418
-as.data.table.array <- function(x, keep.rownames=FALSE, sorted=TRUE, value.name="value", na.rm=TRUE, ...) {
+as.data.table.array <- function(x, keep.rownames=FALSE, key=NULL, sorted=TRUE, value.name="value", na.rm=TRUE, ...) {
```

Comment:
This change is fine, breaking but minor, needs to be mentioned in NEWS. It is quite unlikely that someone was using here argument matching by position on 3+ position.

Reply:
@jangorecki looking again I see now what `sorted` argument does. It's a bit in conflict with potential `key` argument since it's `TRUE` by default.

As of now, user would have to do `as.data.table(ARRAY, key = key_col, sorted = FALSE)`...

Any ideas the right way to proceed? Most natural is to issue an error if `sorted && !is.null(key)`, but it's a bit awkward from user standpoint.

Reply:
maybe just raise error when both are provided? `sorted` is just shortcut for `key=rev(head(names(ans), -1L))`

Reply:
it is unlikely that user will need other key, then the one created by `sorted=TRUE`, on a such structure

Reply:
maybe just mention in doc that `sorted` is ignored if `key` is provided

Reply:
best would be to add some unit tests for those cases too

Reply:
> best would be to add some unit tests for those cases too

was working on this when i thought of the conflict.

> it is unlikely that user will need other key, then the one created by sorted=TRUE, on a such structure

True. Won't bother over-engineering then.